### ISSUE TITLE
Database schema update failed.

### ISSUE THREAD

**Issue by yjeanrenaud** [roundcube/roundcubemail#8014]
I upgraded from 1.3.2 to 1.4.11 and the installer says "Database schema update failed." Schema version: 2016100900 (required: 2019092900). I use php 7.4.16 and MySQL (MariaDB-1:10.3.28) 
I have no access to the installto.sh. I found out there was a similar[ issue related to 1.4.0](https://github.com/roundcube/roundcubemail/issues/7013) but seems to have been fixed.

**Comment by alecpl** [comment]
We'll not be able to help you without some more info, like the error messages from log or command line.

**Comment by yjeanrenaud** [comment]
but there is no installer log and I have no access to others I am afraid.

**Comment by alecpl** [comment]
Then the only thing you can do is to execute the db schema update scripts manually.

**Comment by yjeanrenaud** [comment]
ah I found out my hoster prevented reading the shebang in the scripts correctly. how odd.
anyhow, I now managed to execute installto.sh manually: 
`Copying files to target location...done.

NOTICE: JavaScript dependencies installation skipped.
NOTICE: The 'installer' directory still exists. You should remove it after the upgrade.

Running update script at target...
Executing database schema update.
ERROR: Error in DDL upgrade 2016112200: [1060] Duplicate column name 'id'
Updating database schema (2016112200)... [FAILED]
All done.`

**Comment by yjeanrenaud** [comment]
btw, the same error appears when executing updatedb.sh manually

**Comment by yjeanrenaud** [comment]
ok I fixed it. it was quite messy I am afraid:
```
ALTER TABLE `dictionary` DROP `id`;
```
after that, I got errors on the constraint for the foreign key in creating the table cache. 
```
ERROR: Error in DDL upgrade 2016112200: [1005] Can't create table `jeanrenaud_net`.`cache` (errno: 150 "Foreign key constraint is incorrectly formed")
```
So  I executed the sql statements seperately and found out, the problem layed in the table users. It was on on the wrong engine (MyISAM) and after I changed it, I could also create the foreign key successfully;
```
ALTER TABLE `users` ENGINE=INNODB;
ALTER TABLE `users` CHARACTER SET utf8 COLLATE utf8_general_ci;
```
and then I updated the scheme code manually as well: 
```
UPDATE system SET value = '2016112200' WHERE name = 'roundcube-version';
```
after this, the cli db update ran through without problems:
```
php bin/updatedb.sh --package=roundcube --dir=SQL
Updating database schema (2018021600)... [OK]
Updating database schema (2018122300)... [OK]
Updating database schema (2019092900)... [OK]
```

**Comment by yjeanrenaud** [comment]
ok I fixed it. it was quite messy I am afraid:
```
ALTER TABLE `dictionary` DROP `id`;
```
after that, I got errors on the constraint for the foreign key in creating the table cache. 
```
ERROR: Error in DDL upgrade 2016112200: [1005] Can't create table `hostname`.`cache` (errno: 150 "Foreign key constraint is incorrectly formed")
```
So  I executed the sql statements seperately and found out, the problem layed in the table users. It was on on the wrong engine (MyISAM) and after I changed it, I could also create the foreign key successfully;
```
ALTER TABLE `users` ENGINE=INNODB;
ALTER TABLE `users` CHARACTER SET utf8 COLLATE utf8_general_ci;
```
and then I updated the scheme code manually as well: 
```
UPDATE system SET value = '2016112200' WHERE name = 'roundcube-version';
```
after this, the cli db update ran through without problems:
```
php bin/updatedb.sh --package=roundcube --dir=SQL
Updating database schema (2018021600)... [OK]
Updating database schema (2018122300)... [OK]
Updating database schema (2019092900)... [OK]
```

tldr; I guess it would be wise to check if the engines are selected correctly beforehand. I don't know where the MyISAM came from but it was probably from an older update and it conflicts with the CONSTRAINT statement in 2016112200

**Comment by yjeanrenaud** [comment]
followup: I updated the engine of all roundcube tables now to INNODB and then executed:
```
ALTER TABLE `cache_index` ADD CONSTRAINT `user_id_fk_cache_index` FOREIGN KEY (`user_id`)
   REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `cache_thread` ADD CONSTRAINT `user_id_fk_cache_thread` FOREIGN KEY (`user_id`)
   REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `cache_messages` ADD CONSTRAINT `user_id_fk_cache_messages` FOREIGN KEY (`user_id`)
   REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `contactgroups` ADD CONSTRAINT `user_id_fk_contactgroups` FOREIGN KEY (`user_id`)
    REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
  ALTER TABLE `contactgroupmembers` ADD CONSTRAINT `contactgroup_id_fk_contactgroups` FOREIGN KEY (`contactgroup_id`)
    REFERENCES `contactgroups`(`contactgroup_id`) ON DELETE CASCADE ON UPDATE CASCADE;
  ALTER TABLE `contactgroupmembers` ADD CONSTRAINT `contact_id_fk_contacts` FOREIGN KEY (`contact_id`)
    REFERENCES `contacts`(`contact_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `identities` ADD CONSTRAINT `user_id_fk_identities` FOREIGN KEY (`user_id`)
   REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `dictionary` ADD CONSTRAINT `user_id_fk_dictionary` FOREIGN KEY (`user_id`)
    REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `searches` ADD CONSTRAINT `user_id_fk_searches` FOREIGN KEY (`user_id`) REFERENCES `users`(`user_id`) ON DELETE CASCADE ON UPDATE CASCADE ;
ALTER TABLE `filestore` ADD CONSTRAINT `user_id_fk_filestore` FOREIGN KEY (`user_id`)
   REFERENCES `users` (`user_id`) ON DELETE CASCADE ON UPDATE CASCADE;
 ```

the indexes and UNIQUE statements of these tables were somehow already set to these non-existing foreign keys but nothing complained about that, so I did **not** have to execute
```ALTER TABLE `contacts` ADD  INDEX `user_contacts_index` (`user_id`,`del`); 
ALTER TABLE `searches` ADD UNIQUE `uniqueness` (`user_id`, `type`, `name`);
``` 
etc.  and the table `filestore` already had a propper contraint, too. I suppose we need to implement some more checks in the upgrading routines for propper engines and formatings because otherwise these statements might just be dropped silently due to `/*!40014  SET FOREIGN_KEY_CHECKS=0 */;` a the beginning of _mysql.initial.sql_.
If I am not mistaken and this was not implemented yet, I think I will work on that in the next weeks though and submit a pull request
